### ISSUE TITLE
Toolchain Build Colourisation Lost

### ISSUE THREAD

**Issue by daiyzee** [conan-io/conan#5597]
Conan version 1.18.0
All POSIX plaforms.
CMake build generator.

Our projects are built on about 10 or so different combinations of platform/versions using the Conan build tooling with CMake as the Makefile generator.
On our Ubuntu 16.04 LTS and the debian/stretch/9 systems the terminal colorisation during the build and testing phases has been lost.
I naturally thought it was a change I had made somewhere but then noticed that on our Ubuntu 19.04 build slave all appeared normal.
I turned out that this build slave had been offline and missed the update and was still on 1.17.0. I downgraded the 16.04 LTS and sure enough the colorisation returned.

E.g. The CMake compilation output below would normally be green but is now white as would be displayed on a non-colour supported terminal profile.

[  2%] Building CXX object CMakeFiles/PDUAPI.dir/Users/daiyzee/Projects/gxm/d-pdu-api/src/CTimerCXX11.cpp.o

**Comment by jgsogo** [comment]
Hi @daiyzee, everything related to terminal colorisation is hard and very hard to know what has changed as packages and dependencies may affect it if they are using `colorama` too. I'm afraid I cannot tag this issue as a regression, but we will do our best to get those colors back.

We would need to be able to reproduce the issue to investigate the origin, can you share with us a minimal project that shows that behavior? You can check the python dependencies too, running `pip freeze` will list all the dependencies and versions, if anyone if different that could be a lead too. Thanks!

---
(Somehow related: https://github.com/conan-io/conan/pull/4885)

**Comment by daiyzee** [comment]
Hi, thanks for the response.

To reproduce use a cmake project from conan-center e.g. gtest 1.8.1.
Build it with conan 1.17.0. See the coloured output.
Build it with conan 1.18.X. No coloured output.

It obviously makes it much harder to see the warnings and information from the build pipeline so it really needs to be fixed.

Damon.

**Comment by jgsogo** [comment]
Thanks for the info, @daiyzee .

I see that the change was introduced in 1.18.0 in https://github.com/conan-io/conan/pull/5517, the initialization of `ConanRunner` changed (the implementation is the same):


## Conan <= v1.17.2


Initialization of `ConanRunner` without providing `output` argument
```python
runner = ConanRunner(config.print_commands_to_output, config.generate_run_log_file,                             
                     config.log_run_to_output)
```

it makes Conan use plain `ConanRunner::_simple_os_call` which runs `os.system(command)`

## Conan >= v1.18.0

We are providing the output to the ctor of `ConanRunner`

```python
self.runner = runner or ConanRunner(self.config.print_commands_to_output,
                                    self.config.generate_run_log_file,
                                    self.config.log_run_to_output,
                                    self.out)
```

Conan will use `ConanRunner::_pipe_os_call` which is using `subprocess.Popen` and colors are lost.

---

Not sure if it was intended or not (ping @memsharded)

**Comment by monsdar** [comment]
Same issue over here. Some of our devs wondered about it, but we haven't investigated further yet. Nice to find that it already has been investigated.

This is quite a big UX issue for us, as it is hard to spot any warnings and errors without having the highlighted anymore.

**Comment by monsdar** [comment]
@memsharded: Pinging again. Is this perhaps very easily fixed by simply removing the `self.out` from the call again?

**Comment by memsharded** [comment]
> @memsharded: Pinging again. Is this perhaps very easily fixed by simply removing the self.out from the call again?

Hi! I did try that, but that breaks tons of tests of the suite. So definitely not easy, it will require some time, as we don't have the bandwith to fix it now, sorry.

**Comment by monsdar** [comment]
Got it. Would've been boring if it would just work... Thanks for trying out!

**Comment by monsdar** [comment]
Just quickly checked myself... Not sure what I'm seeing, but that's what I can tell:

* Running `conans/test/functional/build_helpers/cmake_configs_test.py` - `CMakeConfigTests` - `test_package_configs_test`
* With `self.out`: Test is successful
* Without `self.out`: Test is not successful

This does seem to be an issue with the test though: During the test there's a package being created. In the end two executables are executed which simply display the following text:
* `Hello Release Hello0`
* `Hello Debug Hello0`
Somehow this text never reaches cout, resulting in a failing test. The executables however are there and if I change line 20 and 21 from

```
        self.assertIn("Hello Release Hello0", client.out)
        self.assertIn("Hello Debug Hello0", client.out)
```

to

```
        self.assertIn("Hello", client.out)
        self.assertIn("Hello", client.out)
```

the asserts are successful (of course, there's the word Hello somewhere else as well), BUT the output will also be shown in the test output.

To me it looks like Conan is working fine without the `self.out` line, the tests however seem to have hiccups and do not get the whole output... Not sure what to make out of this.

**Comment by monsdar** [comment]
I also checked whether it's possible to display colored text with subprocess.Popen. Unfortunately this seems not to be easy. [Here's a Stackoverflow question](https://stackoverflow.com/questions/34624751/get-output-colored-text-with-popen) that's a good pointer to what's going on.

The SO answer proposing to use pexpect instead of subprocess is not a good option, as it does not seem to support Windows.

**Comment by monsdar** [comment]
Today we discussed this topic again and I again found this old issue. Just asking to make sure - has there been any changes that'd make this fix possible? Is this something on the roadmap anywhere?

**Comment by steffens-a** [comment]
For gcc (version >=4.9) you can use `-fdiagnostics-color=always` as workaround. Add to conan compiler profile
```
[options]
[build_requires]
[env]
CXXFLAGS=-fdiagnostics-color=always
```

**Comment by memsharded** [comment]
Please @czoido include colorization in the CLI 2.0 ongoing work, lets try to make sure it works for build output too if it is possible, and from there lets learn and see if it can be backported.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/7740 fixed this, hopefully everything is good fine.
It will be released in Conan 1.30.
